File: crates/opencorde-core/src/permissions.rs
//! # Permission Bitfield
//! 64-bit permission system using Discord's exact bit layout.
//!
//! Bit values are identical to Discord's permission flag spec so that
//! the client (permissions.ts) and server agree on every flag's meaning.
//! Reference: https://docs.discord.com/developers/topics/permissions
//!
//! ## Depends On
//! - bitflags (efficient bitset type)
//! - serde (JSON serialization)

use crate::snowflake::Snowflake;
use bitflags::bitflags;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

bitflags! {
    /// Permission flags for roles and channels.
    /// Bit positions match Discord's permission flag spec exactly.
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct Permissions: u64 {
        // General server permissions
        const CREATE_INVITE         = 1 << 0;
        const KICK_MEMBERS          = 1 << 1;
        const BAN_MEMBERS           = 1 << 2;
        const ADMINISTRATOR         = 1 << 3;
        const MANAGE_CHANNELS       = 1 << 4;
        const MANAGE_SERVER         = 1 << 5;
        const ADD_REACTIONS         = 1 << 6;
        const VIEW_AUDIT_LOG        = 1 << 7;
        const PRIORITY_SPEAKER      = 1 << 8;
        const STREAM                = 1 << 9;
        const VIEW_CHANNEL          = 1 << 10;
        const SEND_MESSAGES         = 1 << 11;
        const SEND_TTS_MESSAGES     = 1 << 12;
        const MANAGE_MESSAGES       = 1 << 13;
        const EMBED_LINKS           = 1 << 14;
        const ATTACH_FILES          = 1 << 15;
        const READ_MESSAGE_HISTORY  = 1 << 16;
        const MENTION_EVERYONE      = 1 << 17;
        const USE_EXTERNAL_EMOJIS   = 1 << 18;
        const VIEW_GUILD_INSIGHTS   = 1 << 19;
        // Voice permissions
        const CONNECT               = 1 << 20;
        const SPEAK                 = 1 << 21;
        const MUTE_MEMBERS          = 1 << 22;
        const DEAFEN_MEMBERS        = 1 << 23;
        const MOVE_MEMBERS          = 1 << 24;
        const USE_VAD               = 1 << 25;
        // Member management
        const CHANGE_NICKNAME       = 1 << 26;
        const MANAGE_NICKNAMES      = 1 << 27;
        const MANAGE_ROLES          = 1 << 28;
        const MANAGE_WEBHOOKS       = 1 << 29;
        const MANAGE_GUILD_EXPRESSIONS = 1 << 30;
        const USE_APPLICATION_COMMANDS = 1 << 31;
        // Stage / events / threads
        const REQUEST_TO_SPEAK      = 1 << 32;
        const MANAGE_EVENTS         = 1 << 33;
        const MANAGE_THREADS        = 1 << 34;
        const CREATE_PUBLIC_THREADS = 1 << 35;
        const CREATE_PRIVATE_THREADS = 1 << 36;
        const USE_EXTERNAL_STICKERS = 1 << 37;
        const SEND_MESSAGES_IN_THREADS = 1 << 38;
        // Moderation
        const MODERATE_MEMBERS      = 1 << 40;
        const CREATE_EVENTS         = 1 << 44;
        const SEND_VOICE_MESSAGES   = 1 << 46;
        const SEND_POLLS            = 1 << 49;
        const PIN_MESSAGES          = 1 << 51;
        const BYPASS_SLOWMODE       = 1 << 52;
    }
}

impl Permissions {
    /// Default permissions for the @everyone role on a new server.
    pub fn default_everyone() -> Self {
        Permissions::CREATE_INVITE
            | Permissions::VIEW_CHANNEL
            | Permissions::SEND_MESSAGES
            | Permissions::EMBED_LINKS
            | Permissions::ATTACH_FILES
            | Permissions::READ_MESSAGE_HISTORY
            | Permissions::ADD_REACTIONS
            | Permissions::CONNECT
            | Permissions::SPEAK
            | Permissions::USE_VAD
            | Permissions::CHANGE_NICKNAME
    }

    /// All permissions set (used for Administrator bypass).
    pub fn all_permissions() -> Self {
        Permissions::all()
    }
}

impl Serialize for Permissions {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_u64(self.bits())
    }
}

impl<'de> Deserialize<'de> for Permissions {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let bits = u64::deserialize(deserializer)?;
        Ok(Permissions::from_bits_truncate(bits))
    }
}

/// Target type for permission overwrites.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum OverwriteType {
    /// Overwrite applies to a role
    Role,
    /// Overwrite applies to a member
    Member,
}

/// Permission overwrite for a channel.
/// Allows per-channel allow/deny for roles and members.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionOverwrite {
    /// ID of the role or member
    pub id: Snowflake,
    /// Type of target (role or member)
    pub target_type: OverwriteType,
    /// Permissions explicitly allowed
    pub allow: Permissions,
    /// Permissions explicitly denied
    pub deny: Permissions,
}

// Permission computation is in permission_compute module
pub use crate::permission_compute::compute_permissions;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_permissions() {
        let perms = Permissions::default_everyone();
        assert!(perms.contains(Permissions::VIEW_CHANNEL));
        assert!(perms.contains(Permissions::SEND_MESSAGES));
        assert!(perms.contains(Permissions::READ_MESSAGE_HISTORY));
        assert!(perms.contains(Permissions::CONNECT));
        assert!(perms.contains(Permissions::SPEAK));
        assert!(!perms.contains(Permissions::ADMINISTRATOR));
        assert!(!perms.contains(Permissions::MANAGE_CHANNELS));
    }

    #[test]
    fn test_all_permissions() {
        let perms = Permissions::all_permissions();
        assert!(perms.contains(Permissions::ADMINISTRATOR));
        assert!(perms.contains(Permissions::SEND_MESSAGES));
        assert!(perms.contains(Permissions::MANAGE_CHANNELS));
        assert!(perms.contains(Permissions::MANAGE_ROLES));
    }

    #[test]
    fn test_administrator_bit_is_discord_compatible() {
        // ADMINISTRATOR must be bit 3 to match Discord's layout
        assert_eq!(Permissions::ADMINISTRATOR.bits(), 1 << 3);
    }

    #[test]
    fn test_send_messages_bit_is_discord_compatible() {
        assert_eq!(Permissions::SEND_MESSAGES.bits(), 1 << 11);
    }

    #[test]
    fn test_permissions_serialization() {
        let perms = Permissions::SEND_MESSAGES | Permissions::MANAGE_CHANNELS;
        let json = serde_json::to_string(&perms).unwrap();
        let deserialized: Permissions = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized, perms);
    }

    #[test]
    fn test_overwrite_type_serialization() {
        let ow_type = OverwriteType::Role;
        let json = serde_json::to_string(&ow_type).unwrap();
        let deserialized: OverwriteType = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized, ow_type);
    }
}

File: client/src/lib/utils/permissions.ts
/**
 * @file permissions.ts — Permission bit flag definitions
 * @purpose Define all role permission flags with labels, descriptions, and categories
 * @used-by RolePermissionsTab.svelte, permission checks throughout the client
 */

// Bit values match Discord's permission flag spec exactly.
// Reference: https://docs.discord.com/developers/topics/permissions
export const PERMISSIONS = {
  // General server permissions
  CREATE_INSTANT_INVITE:      { bit: 1n << 0n,  label: 'Create Invites',        desc: 'Create invite links',                               category: 'General' },
  KICK_MEMBERS:               { bit: 1n << 1n,  label: 'Kick Members',          desc: 'Remove members from the server',                    category: 'Moderation' },
  BAN_MEMBERS:                { bit: 1n << 2n,  label: 'Ban Members',            desc: 'Permanently ban members',                           category: 'Moderation' },
  ADMINISTRATOR:              { bit: 1n << 3n,  label: 'Administrator',          desc: 'All permissions, bypasses channel overrides',       category: 'General' },
  MANAGE_CHANNELS:            { bit: 1n << 4n,  label: 'Manage Channels',        desc: 'Create, edit, and delete channels',                 category: 'General' },
  MANAGE_GUILD:               { bit: 1n << 5n,  label: 'Manage Server',          desc: 'Change server name, region, icon',                  category: 'General' },
  ADD_REACTIONS:              { bit: 1n << 6n,  label: 'Add Reactions',          desc: 'Add emoji reactions to messages',                   category: 'Text' },
  VIEW_AUDIT_LOG:             { bit: 1n << 7n,  label: 'View Audit Log',         desc: 'View the server audit log',                         category: 'General' },
  PRIORITY_SPEAKER:           { bit: 1n << 8n,  label: 'Priority Speaker',       desc: 'Others\' volume reduced when you speak',            category: 'Voice' },
  STREAM:                     { bit: 1n << 9n,  label: 'Video / Stream',         desc: 'Share camera and screen in voice channels',         category: 'Voice' },
  VIEW_CHANNEL:               { bit: 1n << 10n, label: 'View Channels',          desc: 'View channels and read messages',                   category: 'General' },
  SEND_MESSAGES:              { bit: 1n << 11n, label: 'Send Messages',          desc: 'Send messages in text channels',                    category: 'Text' },
  SEND_TTS_MESSAGES:          { bit: 1n << 12n, label: 'Send TTS Messages',      desc: 'Send text-to-speech messages',                      category: 'Text' },
  MANAGE_MESSAGES:            { bit: 1n << 13n, label: 'Manage Messages',        desc: 'Delete and pin any message',                        category: 'Moderation' },
  EMBED_LINKS:                { bit: 1n << 14n, label: 'Embed Links',            desc: 'Links auto-embed with previews',                    category: 'Text' },
  ATTACH_FILES:               { bit: 1n << 15n, label: 'Attach Files',           desc: 'Upload files and images',                           category: 'Text' },
  READ_MESSAGE_HISTORY:       { bit: 1n << 16n, label: 'Read Message History',   desc: 'Read past messages in channels',                    category: 'Text' },
  MENTION_EVERYONE:           { bit: 1n << 17n, label: 'Mention @everyone',      desc: 'Ping @everyone, @here, and all roles',              category: 'Text' },
  USE_EXTERNAL_EMOJIS:        { bit: 1n << 18n, label: 'Use External Emojis',    desc: 'Use emojis from other servers',                     category: 'Text' },
  VIEW_GUILD_INSIGHTS:        { bit: 1n << 19n, label: 'View Server Insights',   desc: 'View server analytics and insights',                category: 'General' },
  CONNECT:                    { bit: 1n << 20n, label: 'Connect',                desc: 'Join voice and stage channels',                     category: 'Voice' },
  SPEAK:                      { bit: 1n << 21n, label: 'Speak',                  desc: 'Speak in voice channels',                           category: 'Voice' },
  MUTE_MEMBERS:               { bit: 1n << 22n, label: 'Mute Members',           desc: 'Mute others in voice channels',                     category: 'Voice' },
  DEAFEN_MEMBERS:             { bit: 1n << 23n, label: 'Deafen Members',         desc: 'Deafen others in voice channels',                   category: 'Voice' },
  MOVE_MEMBERS:               { bit: 1n << 24n, label: 'Move Members',           desc: 'Move members between voice channels',               category: 'Voice' },
  USE_VAD:                    { bit: 1n << 25n, label: 'Use Voice Activity',      desc: 'Use voice activity detection (not push-to-talk)',   category: 'Voice' },
  CHANGE_NICKNAME:            { bit: 1n << 26n, label: 'Change Nickname',        desc: 'Change own nickname',                               category: 'General' },
  MANAGE_NICKNAMES:           { bit: 1n << 27n, label: 'Manage Nicknames',       desc: "Change other members' nicknames",                   category: 'General' },
  MANAGE_ROLES:               { bit: 1n << 28n, label: 'Manage Roles',           desc: 'Create and edit roles below your highest role',     category: 'General' },
  MANAGE_WEBHOOKS:            { bit: 1n << 29n, label: 'Manage Webhooks',        desc: 'Create and manage webhooks',                        category: 'General' },
  MANAGE_GUILD_EXPRESSIONS:   { bit: 1n << 30n, label: 'Manage Emojis & Stickers', desc: 'Add, edit, and remove custom emojis/stickers',   category: 'General' },
  USE_APPLICATION_COMMANDS:   { bit: 1n << 31n, label: 'Use Slash Commands',     desc: 'Use application slash commands and context menus', category: 'Text' },
  REQUEST_TO_SPEAK:           { bit: 1n << 32n, label: 'Request to Speak',       desc: 'Request to speak in stage channels',                category: 'Stage' },
  MANAGE_EVENTS:              { bit: 1n << 33n, label: 'Manage Events',          desc: 'Edit and delete scheduled events',                  category: 'Events' },
  MANAGE_THREADS:             { bit: 1n << 34n, label: 'Manage Threads',         desc: 'Archive and delete threads',                        category: 'Text' },
  CREATE_PUBLIC_THREADS:      { bit: 1n << 35n, label: 'Create Public Threads',  desc: 'Create public threads on messages',                 category: 'Text' },
  CREATE_PRIVATE_THREADS:     { bit: 1n << 36n, label: 'Create Private Threads', desc: 'Create invite-only private threads',                category: 'Text' },
  USE_EXTERNAL_STICKERS:      { bit: 1n << 37n, label: 'Use External Stickers',  desc: 'Use stickers from other servers',                   category: 'Text' },
  SEND_MESSAGES_IN_THREADS:   { bit: 1n << 38n, label: 'Send in Threads',        desc: 'Send messages in threads',                          category: 'Text' },
  MODERATE_MEMBERS:           { bit: 1n << 40n, label: 'Timeout Members',        desc: 'Temporarily mute members (timeout)',                category: 'Moderation' },
  CREATE_EVENTS:              { bit: 1n << 44n, label: 'Create Events',          desc: 'Create scheduled events',                           category: 'Events' },
  SEND_VOICE_MESSAGES:        { bit: 1n << 46n, label: 'Send Voice Messages',    desc: 'Send voice message recordings',                     category: 'Text' },
  SEND_POLLS:                 { bit: 1n << 49n, label: 'Send Polls',             desc: 'Create poll messages',                              category: 'Text' },
  PIN_MESSAGES:               { bit: 1n << 51n, label: 'Pin Messages',           desc: 'Pin and unpin messages in channels',                category: 'Moderation' },
  BYPASS_SLOWMODE:            { bit: 1n << 52n, label: 'Bypass Slowmode',        desc: 'Send messages ignoring slowmode restrictions',      category: 'Moderation' },
} as const;

export type PermissionKey = keyof typeof PERMISSIONS;

export const PERMISSION_CATEGORIES = ['General', 'Moderation', 'Text', 'Voice', 'Stage', 'Events'] as const;
export type PermissionCategory = (typeof PERMISSION_CATEGORIES)[number];

/** Check whether a given flag is set in a permissions bigint. */
export function hasPermission(permissions: bigint, flag: bigint): boolean {
  return (permissions & flag) === flag;
}

/** Return a new permissions bigint with the flag toggled on or off. */
export function togglePermission(permissions: bigint, flag: bigint, enabled: boolean): bigint {
  return enabled ? (permissions | flag) : (permissions & ~flag);
}

/** Convert bigint permissions to a number for JSON serialisation (API uses i64/u64). */
export function permissionsToNumber(perms: bigint): number {
  return Number(perms & 0xFFFFFFFFFFFFFFFFn);
}

/** Convert a number from the API into a client-side bigint. */
export function numberToPermissions(n: number): bigint {
  return BigInt(n);
}
